### PR TITLE
TH2-2036. Message filtering by timestamp is inefficient

### PR DESCRIPTION
**Pull request by MrJonniK**

### CODE REVIEW

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/messages/MessageBatchQueryProvider.java [hunk 1]
```diff
@@ -224,24 +218,56 @@ private BoundStatementBuilder bindFilterParameters(BoundStatementBuilder builder
...
-			builder = FilterUtils.bindTimestamp(ts.plus(10, ChronoUnit.MINUTES), builder, LAST_MESSAGE_DATE, LAST_MESSAGE_TIME);
+			try
+			{
+				long rightBatchIndex = getNearestMessageIndexAfter(tmOperator, instanceId,
+						filter.getStreamName().getValue(), directionFilter.getValue(), ts, attributes);
+				builder = builder.setLong(RIGHT_MESSAGE_INDEX, rightBatchIndex);
+			}
+			catch (ExecutionException | InterruptedException e)
+			{
+				throw new CradleStorageException("Error getting message batch index for timestamp 'To'", e);
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/messages/MessageBatchQueryProvider.java:247]: throw new CradleStorageException("Error while getting message batch index for timestamp 'To'="+ts, e);

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/messages/MessageBatchQueryProvider.java [hunk 2]
```diff
@@ -224,24 +218,56 @@ private BoundStatementBuilder bindFilterParameters(BoundStatementBuilder builder
...
-		//FIXME: This doesn't guarantee that batches longer than 10 minutes will be not skipped! Anyway, such batches are bad.
 		if (filter.getTimestampFrom() != null)
 		{
 			Instant ts = filter.getTimestampFrom().getValue();
-			builder = FilterUtils.bindTimestamp(ts.minus(10, ChronoUnit.MINUTES), builder, FIRST_MESSAGE_DATE, FIRST_MESSAGE_TIME);
+			try
+			{
+				long leftBatchIndex = getNearestMessageIndexBefore(tmOperator, instanceId,
+						filter.getStreamName().getValue(), directionFilter.getValue(), ts, attributes);
+				builder = builder.setLong(LEFT_MESSAGE_INDEX, leftBatchIndex);
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/messages/MessageBatchQueryProvider.java:228]: There seem to be a conflict if someone defines both timestampFrom and index in the filter. Better use different placeholder name for index obtained for timestampFrom.

**Review by Vladimir-Panarin-Exactpro — CHANGES_REQUESTED** [review]

**Review by Vladimir-Panarin-Exactpro — APPROVED** [review]

**Comment by MrJonniK** [comment]
This commit doesn't work if the filter contains index and from/to constraints. Cassandra throw the next exception `More than one restriction was found for the start bound  on message_index`

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/messages/MessageBatchQueryProvider.java [hunk 3]
```diff
@@ -104,65 +106,76 @@ private Select orderBy(Order order, Select select)
...
+						select.whereColumn(MESSAGE_INDEX), LEFT_MESSAGE_INDEX);
+				select = FilterUtils.filterToWhere(ComparisonOperation.LESS_OR_EQUALS,
+						select.whereColumn(MESSAGE_INDEX), RIGHT_MESSAGE_INDEX);
+				return select;
+			}
+			
 			//This is for case when need to return "previous X messages, i.e. X messages whose index is less than Y"
-			if (filter.getLimit() > 0 && (operation == ComparisonOperation.LESS || operation == ComparisonOperation.LESS_OR_EQUALS))
-				select = FilterUtils.filterToWhere(ComparisonOperation.GREATER_OR_EQUALS, select.whereColumn(MESSAGE_INDEX), LEFT_MESSAGE_INDEX);
+			if (filter.getLimit() > 0 && (op == ComparisonOperation.LESS || op == ComparisonOperation.LESS_OR_EQUALS))
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/messages/MessageBatchQueryProvider.java:132]: `else if`

**Review by Vladimir-Panarin-Exactpro — CHANGES_REQUESTED** [review]

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/messages/MessageBatchQueryProvider.java (hunk 3 above)
Comment (MrJonniK) [reply cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/messages/MessageBatchQueryProvider.java:132]: But if above contains return statement

Comment (Vladimir-Panarin-Exactpro) [reply cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/messages/MessageBatchQueryProvider.java:132]: Ok.

**Review by Vladimir-Panarin-Exactpro — APPROVED** [review]

**Pull request merged**